### ISSUE TITLE
Do we need to compensate for the 3d box annotation when we project it to the image?

### ISSUE THREAD

**Issue by xmyqsh** [argoai/argoverse-api#90]
I noticed that when we project lidar to the image, we should compensate for the motion due to the unperfect time sync which is done in `project_lidar_to_img_motion_compensated`.

I think we should also need to compensate for the 3d box annotation when we project it to the image. Because the annotation timestamp is same as the lidar sweep. But `project_ego_to_cam` does not compensate it.

**Comment by johnwlambert** [comment]
Hello @xmyqsh, thanks for your interest in our work. 

As you noted, we provide a function `project_lidar_to_img_motion_compensated()` to project LiDAR points into an image (and compensating for the egomotion between the two sensor observations) [here](https://github.com/argoai/argoverse-api/blob/master/demo_usage/cuboids_to_bboxes.py#L179
). The annotations are present in the LiDAR sweeps, so we only need to consider compensating between LiDAR and images.

This function requires providing the timestamp of both sensor observations:
```python
                    if motion_compensate:
                        uv, uv_cam, valid_pts_bool, K = project_lidar_to_img_motion_compensated(
                            points_h,  # these are recorded at lidar_time
                            copy.deepcopy(log_calib_data),
                            camera_name,
                            cam_timestamp,
                            lidar_timestamp,
                            data_dir,
                            log_id,
                            return_K=True,
                        )
```

The [function](https://github.com/argoai/argoverse-api/blob/652bb500eafd2b21a8a346a3b762c1114f4c314a/argoverse/utils/calibration.py#L665
) works by moving LiDAR points in the egovehicle frame at time `t0` into the egovehicle frame at time `t1` when the image was recorded, using the map as a static reference frame to unify the two.

```python
    ego_cam_t_SE3_ego_lidar_t = city_SE3_ego_cam_t.inverse().right_multiply_with_se3(city_SE3_ego_lidar_t)
    pts_h_cam_time = ego_cam_t_SE3_ego_lidar_t.transform_point_cloud(pts_h_lidar_time)
```

As you noted, we also provide a version of such a projection function without any motion compensation, simply for the sake of illustration and completion ([here](https://github.com/argoai/argoverse-api/blob/652bb500eafd2b21a8a346a3b762c1114f4c314a/argoverse/utils/calibration.py#L132)). I would not recommend using `project_ego_to_cam` (and we do not use it in our demo file's [default settings](https://github.com/argoai/argoverse-api/blob/652bb500eafd2b21a8a346a3b762c1114f4c314a/demo_usage/cuboids_to_bboxes.py#L101) either.

**Comment by xmyqsh** [comment]
@johnwlambert 
Thanks for your reply. As you said that the annotations are present in the LiDAR sweeps, I think
it is also necessary to consider compensating between 3d annotations and images. Right?

But it seems that some of the annotations are not very accurate, no matter I compensated it or not.

```
root_dir =  '../argoverse-tracking/train1/'
log_id = 'c6911883-1843-3727-8eaa-41dc8cda8993'
idx = 0
```

![image](https://user-images.githubusercontent.com/16010828/69335438-efbda680-0c97-11ea-9d43-aed4dd1e5729.png)
![image](https://user-images.githubusercontent.com/16010828/69335457-f8ae7800-0c97-11ea-948f-eaff8fa5c47b.png)
![image](https://user-images.githubusercontent.com/16010828/69336348-411a6580-0c99-11ea-9648-6146e9e9e4c5.png)
![image](https://user-images.githubusercontent.com/16010828/69336357-4aa3cd80-0c99-11ea-92cb-5c8281704cac.png)




________________________________________________________________________________________________
While, when `idx = 150` it looks very well:
![image](https://user-images.githubusercontent.com/16010828/69336009-d9fcb100-0c98-11ea-8786-09e788bc94cd.png)
![image](https://user-images.githubusercontent.com/16010828/69336019-e1bc5580-0c98-11ea-9774-745ffc6716d8.png)
![image](https://user-images.githubusercontent.com/16010828/69336030-ea149080-0c98-11ea-9af8-b97dd449d2d5.png)

Can you help to have a double-check?

**Comment by johnwlambert** [comment]
Hi @xmyqsh, I took a look at the video and it looks like either the LiDAR or the camera starts recording momentarily before the other, so the LiDAR annotation seems to stand still for the first, second, and third image, before we start seeing full sensor observations kick in.  This leads to a misalignment of ~5-10 cm (not noticeable on a car or person, but only on a bollard) for the first two images.

The LiDAR observation occurs once per 100 ms and the image observation occurs once per ~33 ms.

As far as I can tell, the LiDAR-camera alignment/synchronization is correct for every single other object (hundreds of other bounding boxes) in the video except for those two skinny bollards in the first frame.

I would recommend discarding the first LiDAR frame of this sequence if these bollards present an issue for you.

I've uploaded the relevant videos here:
https://youtu.be/XlRpSpFL5ZI

https://youtu.be/Qnb2AgDHozI
